### PR TITLE
Docs and unit tests for app_server.rs

### PR DESCRIPTION
**Pull request by koraa**

### CODE REVIEW

**Review by aparcar — CHANGES_REQUESTED** [review]

File: rosenpass/src/protocol/protocol.rs [hunk 1]
```diff
@@ -2899,7 +2899,7 @@ impl CryptoServer {
     /// This is a lot of code. If you want to read the file outside of the documentation,
     /// check out `rosenpass/tests/poll_example.rs" in the repository.
     ///
-    #[doc = "```"]
+    #[doc = "```ignore"]
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/protocol/protocol.rs:2902]: Why? there is no reason to compile twice.

File: rosenpass/src/app_server.rs [hunk 2]
```diff
@@ -660,22 +941,37 @@ impl AppServer {
         })
     }
 
+    /// Access the cryptographic protocol server
+    ///
+    /// This may return an error if [Self] was initialized without a keypair
+    /// and no keypair has been supplied since then.
+    ///
+    /// I.e. will return an error if [Self::crypto_site] is not fully erect
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/app_server.rs:949]: Can you push something better?

File: rosenpass/src/app_server.rs [hunk 3]
```diff
@@ -470,18 +742,22 @@ impl HostPathDiscoveryEndpoint {
...
         })
     }
 
+    /// List of address candidates for the peer
     pub fn addresses(&self) -> &Vec<SocketAddr> {
         &self.addresses
     }
 
+    /// Calculates and stores the next value for [Self::scouting_state]
+    /// given the address and socket we just send a scouting [crate::msgs::InitHello] message
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/app_server.rs:759]: Can you push the fix?

File: rosenpass/src/app_server.rs [hunk 4]
```diff
@@ -406,6 +665,8 @@ impl Endpoint {
         Some(Self::discovery_from_addresses(addrs))
     }
 
+    /// Send a message to the address referenced to by this endpoint or to one of
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/app_server.rs:668]: Can you push the fix?

File: rosenpass/src/app_server.rs [hunk 5]
```diff
@@ -299,14 +538,21 @@ impl std::fmt::Display for Endpoint {
...
     /// determined when we actually receive an RespHello message
     socket: SocketPtr,
-    /// Just the address
+    /// The network address
     addr: SocketAddr,
-    /// identifier
+    /// Byte representation of this socket bound network address.
+    /// Generated through [SocketBoundEndpoint::to_bytes].
+    ///
+    /// Read though [HostIdentification::encode]
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/app_server.rs:555]: Can you push the fix?

File: rosenpass/src/app_server.rs [hunk 6]
```diff
@@ -201,43 +405,61 @@ pub struct AppServer {
...
+    /// Retrieve the [AppPeer] pointed to by [Self], mutably
     pub fn get_app_mut<'a>(&self, srv: &'a mut AppServer) -> &'a mut AppPeer {
         &mut srv.peers[self.0]
     }
 
+    /// Use the associated WireGuard PSK broker via [BrokerStorePtr]
+    /// to upload a new PSK.
+    ///
+    /// If no PSK broker is set and [AppPeer::outfile] is none, then
+    /// this prints a warning
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/app_server.rs:462]: Thanks for pointing this out, can you open a ticket?

File: rosenpass/src/app_server.rs [hunk 7]
```diff
@@ -151,43 +298,100 @@ pub struct AppServerTest {
...
+/// This holds pretty much all of the state of the Rosenpass application
+/// including the cryptographic state in [Self::crypto_site]
+///
+/// Responsible for file IO, network IO, etc…
 ///
-/// Responsible for file IO, network IO
-// TODO add user control via unix domain socket and stdin/stdout
+/// # Examples
+///
+#[doc = "```ignore"]
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/app_server.rs:326]: It is not just compiled; it is executed during the regular tests. This is a test that simply lives in the integration test directory.

File: rosenpass/src/app_server.rs [hunk 8]
```diff
@@ -49,83 +51,220 @@ use crate::{
...
+    /// use rosenpass_util::functional::run;
+    ///
+    /// let mut peer = AppPeer {
+    ///   outfile: None,
+    ///   broker_peer: None,
+    ///   initial_endpoint: Some(Endpoint::discovery_from_hostname("0.0.0.0:0".to_string())?),
+    ///   current_endpoint: Some(Endpoint::discovery_from_hostname("0.0.0.0:1".to_string())?),
+    /// };
+    ///
+    /// fn same(fst: Option<&Endpoint>, snd: Option<&Endpoint>) -> bool {
```
[the remark this answers is not in the document]
Comment (koraa) [reply rosenpass/src/app_server.rs:234]: Can you improve the code accordingly?

**Review by aparcar — APPROVED** [review]

**Pull request merged**